File: Codes/C++/search_algorithms/binary_search.cpp
#include <iostream>

using namespace std;

// size of array
int N = 10;

int binary_search(int arr[], int low, int up, int x)
{
    while (low <= up) {
        int middle = low + (up - low) / 2;

        // Check if x is present at mid
        if (arr[middle] == x)
            return middle;

        // If x greater, ignore left half
        if (arr[middle] < x)
            low = middle + 1;

        // If x is smaller, ignore right half
        else
            up = middle - 1;
    }

    // if we reach here, then element was
    // not present
    return -1;
}


int main(){
	int arr[N] = {1,2,5,25,35,67,87,88,90,100};


	for(auto e: arr){ cout << e << " " ;} 
	// input the element
	cout << endl;
	cout << "Enter the element you want to search for: " ;
	int element; cin >> element;
	int found =	binary_search(arr,0,N-1,element);
	found!=-1 ? cout << "Element exists" << endl : cout << "Element does not exist in the array" << endl;
}

File: Codes/Python/Searching_algorithms/binary_search.py
# It returns index of n in given array arr if present,
# else returns -1
def binary_search(arr, n):
    low = 0
    high = len(arr) - 1
    mid = 0

    while low <= high:

        mid = (high + low) // 2

        # Check if n is present at mid-pos
        if arr[mid] < n:
            low = mid + 1

        # If n is greater, remove left half
        elif arr[mid] > n:
            high = mid - 1

        # If n is smaller, remove right half
        else:
            return mid

        # If the loop ends, then the element was not found
    return -1


# Test array
arr = [4,4,34,35,4,324,6,4,3,43,342]
n = 342

result = binary_search(arr, n)
print(result)